### PR TITLE
Extract components

### PR DESCRIPTION
**Pull request by jniechcial**
New PR due to rebasing and fixing conflicts, but the discussion [available in this PR](https://github.com/JedWatson/react-select/pull/328).

### CODE REVIEW

**Comment by bruderstein** [comment]
Cool, I've just rebased this on my even-more-tests branch (this will be sent as a PR if/when PR #348 is merged, as it relies on it), and one test fails. (The tests for valueRenderer are sadly only in this branch)

It's a pretty easy fix though, you just need to move the 
```
var val = this.state.values[0] || null;
```
from line 767 to outside the deepest if, otherwise in the case of a single value with a valueRenderer, you pass option={val}, but val isn't defined.  After doing that, the test passes, and everything is green.

Tests are nice :)

File: src/Select.js [hunk 1]
```diff
@@ -720,29 +739,39 @@ var Select = React.createClass({
...
 			if(this.props.valueRenderer && !!this.state.values.length) {
-				var val = this.state.values[0] || null;
 				value.push(<Value
 						key={0}
 						option={val}
 						renderer={this.props.valueRenderer}
 						disabled={this.props.disabled} />);
 			} else {
-				value.push(<div className="Select-placeholder" key="placeholder">{this.state.placeholder}</div>);
+				var val = this.state.values[0] || null;
```
Comment (bruderstein) [inline src/Select.js]: Need to move this line up

File: src/Select.js [hunk 2]
```diff
@@ -720,29 +739,39 @@ var Select = React.createClass({
...
+					onRemove: onRemove,
+					disabled: this.props.disabled
+				});
+
+				value.push(valueComponent);
 			}, this);
 		}
 
 		if(!this.state.inputValue && (!this.props.multi || !value.length)) {
 			if(this.props.valueRenderer && !!this.state.values.length) {
```
Comment (bruderstein) [inline src/Select.js]: line 767 should be here

File: src/Select.js [hunk 3]
```diff
@@ -720,29 +739,39 @@ var Select = React.createClass({
...
-				var val = this.state.values[0] || null;
 				value.push(<Value
 						key={0}
 						option={val}
```
Comment (bruderstein) [inline src/Select.js]: val used here, without being defined

**Comment by jniechcial** [comment]
@bruderstein fixed, good to have wide test coverage :)

**Comment by JedWatson** [comment]
#348 is merged :) this looks great @jniechcial, thank you. and thanks to @bruderstein for the review!

**Pull request merged**